### DISCUSSION TITLE
Server-side rendering not working on production in Next.js@12.1.6

### DISCUSSION THREAD

**Discussion by addy-pathania** [vercel/next.js#37826] [Help] [answered]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

 Operating System:
      Platform: Windows
      Arch: x64
      Version: 20H2
    Binaries:
      Node: v14.17.6
      npm: 8.1.3
      Yarn: 1.22.17
    Relevant packages:
      next: 12.1.6
      react: 17.0.2
      react-dom: 17.0.2

### What browser are you using? (if relevant)

Chrome 102.0.5005.115

### How are you deploying your application? (if relevant)

next build

### Describe the Bug

When using server-side rendering with next@12.1.6, the context doesn't get passed down to the children component.
Due to this, the ReactDom does not work as expected and the expected component is not getting rendered on the page.

NOTE: This issue cannot be reproduced with other versions of nextjs its only specific to 12.1.6.
           The issue also gets resolved when upgrading react to the latest version i.e. 18.2.0 (this also works with next@12.1.6)
           Due to breaking changes we cannot upgrade to react18 and asking our customer to downgrade to lower versions of nextjs 
           doesn't sound very friendly.

### Expected Behavior

The context should get passed down to the children and all components should render as expected.

### Link to reproduction

N/A

### To Reproduce

Described above

**Comment by icyJoseph** [comment]
Hi,

Do you have a small reproduction demo?

**Reply by im-ghost** [reply]
I tried using express custom server for my next project based and the example given,but I get an error saying next is not a function

**Reply by addy-pathania** [reply]
Hi @icyJoseph  thanks for replying. 
Please try to run the attached nextjs app files.
Steps: `next build` --> `next start`  --> Open the `Demo` component. Props are coming undefined on server.

[my-app.zip](https://github.com/vercel/next.js/files/8994496/my-app.zip)

**Reply by icyJoseph** [reply]
Please put it on a GitHub repo.

**Reply by addy-pathania** [reply]
@icyJoseph https://github.com/sc-addypathania/nextjs-bug-repro/tree/main

**Comment by icyJoseph** [comment, marked as answer]
Hi, @sc-addypathania I think there's a valid case for a bug here. 

The production build misses the context, but only for the situation you have set-up in: https://github.com/sc-addypathania/nextjs-bug-repro/tree/main

I have **confirmed** that `useContext` works fine, and also `contextType` works fine as well.

### Quick Options

#### Refactor the `HoC`

```tsx
export function withComponentFactory(Component) {
  function WithComponentFactory(props) {
    const context = useContext(StateContext);

    return <Component {...props} context={context} />;
  }

  WithComponentFactory.displayName = `withComponentFactory(${
    Component.displayName || Component.name || "Anonymous"
  })`;

  return WithComponentFactory;
}
```

#### use `contextType`

```tsx
import { StateContext } from "../context/GlobalState";
import React from "react";

export class Demo extends React.Component {
  render() {
    console.log(this.props);

    return <h2>Hello</h2>;
  }
}

export class Demo1 extends React.Component {
  render() {
    console.log(this);

    return <Demo {...this.props} context={this.context} />;
  }
}

Demo1.contextType = StateContext;
```

### With `useContext`

```tsx
import { StateContext } from "../context/GlobalState";
import React, { useContext } from "react";

export class Demo extends React.Component {
  render() {
    console.log(this.props);

    return <h2>Hello</h2>;
  }
}

export const Demo1 = (props) => {
  const context = useContext(StateContext);

  return <Demo {...props} context={context} />;
};
```

These are counter measures to get around the issue.

I think you should re-open this as a bug, with the reproduction demo, and the snippets for `GlobalState.js` - also describe clearly that this works with `next dev` but fails with `next build && next start`.

**Reply by addy-pathania** [reply]
@icyJoseph thanks for the suggestions. I will give them a try. :)
